PhysicsPackages/IMCPhysicsPackage_class.f90

```diff
@@ -0,0 +1,507 @@
+
+  contains
+    procedure :: init
+    procedure :: printSettings
+    procedure :: cycles
+    procedure :: collectResults
+    procedure :: run
+    procedure :: kill
+
+  end type IMCPhysicsPackage
+
+contains
+
+  subroutine run(self)
+    class(IMCPhysicsPackage), intent(inout) :: self
+
+    print *, repeat("<>",50)
+    print *, "/\/\ IMC CALCULATION /\/\"
+
+    call self % cycles(self % tally, self % imcWeightAtch, self % N_cycles)
+    call self % collectResults()
+
+    print *
+    print *, "\/\/ END OF IMC CALCULATION \/\/"
+    print *
+  end subroutine
+
+  !!
+  !! Run cycles for calculation
+  !!
+  subroutine cycles(self, tally, tallyAtch, N_cycles)
+    class(IMCPhysicsPackage), intent(inout)         :: self
+    type(tallyAdmin), pointer,intent(inout)         :: tally
+    type(tallyAdmin), pointer,intent(inout)         :: tallyAtch
+    integer(shortInt), intent(in)                   :: N_cycles
+    integer(shortInt)                               :: i, j, N
+    type(particle)                                  :: p
+    real(defReal)                                   :: elapsed_T, end_T, T_toEnd, sumT
+    real(defReal), dimension(:), allocatable        :: tallyEnergy
+    class(IMCMaterial), pointer                     :: mat
+    logical(defBool)                                :: printUpdates
+    character(100),parameter :: Here ='cycles (IMCPhysicsPackage_class.f90)'
+    class(tallyResult), allocatable                 :: tallyRes
+
+    ! Set whether or not to print energy and temperature updates of each material
+    !   Printed from updateMat (baseMgIMCMaterial_class.f90), 7 lines of text
+    !   per material so recommend to only print when low number of materials
+    if (self % nMat <= 5) then
+      printUpdates = .True.
+    else
+      printUpdates = .False.
+    end if
+
+    ! Attach nuclear data and RNG to particle
+    p % pRNG   => self % pRNG
+    p % timeMax = self % deltaT
+    p % geomIdx = self % geomIdx
+
+    ! Reset and start timer
+    call timerReset(self % timerMain)
+    call timerStart(self % timerMain)
+
+    allocate(tallyEnergy(self % nMat))
+
+    do i=1,N_cycles
+
+      ! Store photons remaining from previous cycle
+      self % thisCycle = self % nextCycle
+      call self % nextCycle % cleanPop()
+
+      ! Check that there are regions of non-zero temperature by summing mat temperatures
+      sumT = 0
+      do j=1, self % nMat
+        mat => IMCMaterial_CptrCast(self % nucData % getMaterial(j))
+        sumT = sumT + mat % getTemp()
+      end do
+
+      N = self % pop
+
+      ! Generate IMC source, only if there are regions with non-zero temperature
+      if(sumT > 0) then
+        ! Select number of particles to generate
+        if(N + self % thisCycle % getSize() > self % limit) then
+          ! Fleck and Cummings IMC Paper, eqn 4.11
+          N = self % limit - self % thisCycle % getSize() - self % nMat - 1
+        end if
+        if(self % sourceGiven) N = N/2
```

Comment:
I have no idea myself - is this a common heuristic? Why should half of particles be from the source?